File: solutions/Java/copy_list_with_random_pointer_hash_map.java
import java.util.HashMap;
import java.util.Map;

class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}

class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) return null;

        // Map original nodes to their copies
        Map<Node, Node> nodeMap = new HashMap<>();

        // First pass: create all copy nodes
        Node current = head;
        while (current != null) {
            nodeMap.put(current, new Node(current.val));
            current = current.next;
        }

        // Second pass: set next and random pointers
        current = head;
        while (current != null) {
            Node copyNode = nodeMap.get(current);
            copyNode.next = current.next != null ? nodeMap.get(current.next) : null;
            copyNode.random = current.random != null ? nodeMap.get(current.random) : null;
            current = current.next;
        }

        return nodeMap.get(head);
    }
}

File: solutions/C++/copy_list_with_random_pointer_hash_map.cpp
#include <unordered_map>
using namespace std;

class Node {
public:
    int val;
    Node* next;
    Node* random;

    Node(int _val) : val(_val), next(NULL), random(NULL) {}
};

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if (!head) return NULL;

        // Map original nodes to their copies
        unordered_map<Node*, Node*> nodeMap;

        // First pass: create all copy nodes
        Node* current = head;
        while (current) {
            nodeMap[current] = new Node(current->val);
            current = current->next;
        }

        // Second pass: set next and random pointers
        current = head;
        while (current) {
            Node* copyNode = nodeMap[current];
            copyNode->next = current->next ? nodeMap[current->next] : NULL;
            copyNode->random = current->random ? nodeMap[current->random] : NULL;
            current = current->next;
        }

        return nodeMap[head];
    }
};
